lib/server.js

```diff
@@ -0,0 +1,222 @@
+'use strict'
+
+const path = require('path')
+const Joi = require('joi')
+const Camp = require('camp')
+const makeBadge = require('../gh-badges/lib/make-badge')
+const GithubConstellation = require('../services/github/github-constellation')
+const { loadServiceClasses } = require('../services')
+const analytics = require('./analytics')
+const { makeBadgeData } = require('./badge-data')
+const log = require('./log')
+const { staticBadgeUrl } = require('./make-badge-url')
+const suggest = require('./suggest')
+const sysMonitor = require('./sys/monitor')
+const PrometheusMetrics = require('./sys/prometheus-metrics')
+const { makeSend } = require('./result-sender')
+const { handleRequest, clearRequestCache } = require('./request-handler')
+const { clearRegularUpdateCache } = require('./regular-update')
+
+const optionalUrl = Joi.string().uri({ scheme: ['http', 'https'] })
+const requiredUrl = optionalUrl.required()
+
+const configSchema = Joi.object({
+  bind: {
+    port: Joi.number()
+      .port()
+      .required(),
+    address: Joi.string()
+      .ip()
+      .required(),
+  },
+  metrics: {
+    prometheus: {
+      enabled: Joi.boolean().required(),
+      allowedIps: Joi.string(),
+    },
+  },
+  ssl: {
+    isSecure: Joi.boolean().required(),
+    key: Joi.string(),
+    cert: Joi.string(),
+  },
+  baseUri: requiredUrl,
+  redirectUri: optionalUrl,
+  cors: {
+    allowedOrigin: Joi.array().items(Joi.string().hostname()),
+  },
+  persistence: {
+    dir: Joi.string().required(),
+    redisUrl: optionalUrl,
+  },
+  services: {
+    github: {
+      baseUri: requiredUrl,
+      debug: {
+        enabled: Joi.boolean().required(),
+        intervalSeconds: Joi.number()
+          .integer()
+          .min(1),
+      },
+    },
+    trace: Joi.boolean().required(),
+  },
+  profiling: {
+    makeBadge: Joi.boolean().required(),
+  },
+  cacheHeaders: {
+    defaultCacheLengthSeconds: Joi.number().integer(),
+  },
+  rateLimit: Joi.boolean().required(),
+  handleInternalErrors: Joi.boolean().required(),
+}).required()
+
+module.exports = class Server {
```

Comment:
Basically this is a rewrite of `server.js` as an object which as much as possible encapsulates its own state.